feat(helm): add --use-source-hooks flag to helm 'rollback'

JPula opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
Provides option to use rollback hooks from current release instead of target release when running helm 'rollback'

**Special notes for your reviewer**:
Added the flag to remain backwards compatible.
closes #12757

**If applicable**:
- [X] this PR contains documentation
- [X] this PR contains unit tests
- [X] this PR has been tested for backwards compatibility

JPula committed:
feat(helm): add --use-source-hooks flag to helm 'rollback'

Signed-off-by: Jared Stehr Rivera <jared.s.rivera@gmail.com>

z4ce commented:
I'd like to see a test that would fail if the wrong hooks were run. Otherwise, looks good to me

JPula committed:
add tests for hook execution during rollback

Signed-off-by: Jared Stehr Rivera <jared.s.rivera@gmail.com>

JPula commented:
@z4ce Thanks for the feedback! 
I have added 2 tests to check whether source or target hooks were run during rollback. Cheers!

(I forgot to signoff my last commit 🤦‍ so I rebased my commits, apologies!)

mattfarina commented:
@JPula thanks for the PR. Can you please fix the linting issue.

JPula committed:
fix linting issues

Signed-off-by: Jared Stehr Rivera <jared.s.rivera@gmail.com>

JPula commented:
@mattfarina I have fixed the linting issue

Gin-sen commented:
Hi, I was wondering how would that work with the "helm upgrade --atomic" command ?

Can we add the "--use-source-hooks" to it so the rollback triggered by a failed upgrade still trigger the sources hooks ?

Very glad to see this feature been addressed, thx a lot !

JPula commented:
> Hi, I was wondering how would that work with the "helm upgrade --atomic" command ?
> 
> Can we add the "--use-source-hooks" to it so the rollback triggered by a failed upgrade still trigger the sources hooks ?
> 
> Very glad to see this feature been addressed, thx a lot !
  
Adding the behavior for atomic upgrades wouldn't be difficult. Though the flag name should be more explicit if it is added to the 'helm upgrade' command (maybe "--use-source-hooks-on-rollback" ?).

That said, I think it would be better if you create a new ticket for this.
For now this PR only aims to add the flag to the helm 'rollback' command.

mluckam commented:
@JPula thanks for the PR, great feature add!  @z4ce and @mattfarina what are next steps to ensure this gets merged?  Looks like it is close.  Glad to help out however possible.

z4ce commented:
@gjenkins8 @scottrigby might be a fresh set of eyes for this one

mluckam commented:
@z4ce thanks for the quick response.   @mattfarina @gjenkins8 @scottrigby mind taking a look at this PR?  Linting issues were addressed as requested Feb 10, 2024.

balmeida-nokia commented:
@mattfarina @gjenkins8 @scottrigby, any news?

balmeida-nokia commented:
@mattfarina Is this on track for 3.19?
@JPula Seems to need a rebase.

balmeida-nokia commented:
@JPula Seems to need a rebase.

vfedorishchev commented:
Hi 👋 is there any help needed to make this PR ready for merge? I think it's a pretty good addition, allowing for auto-rollback execution for DB migrations.
